### ISSUE TITLE
Qwen3-4B base greedy drift vs HF bf16 due to RoPE bf16 materialization boundary

### ISSUE THREAD

**Issue by NolanHo** [xiaguan/pegainfer#174]
## Summary

Qwen3-4B base greedy generation can diverge from Hugging Face GPU bf16 on tie-sensitive logits because PegaInfer previously missed several bf16 materialization boundaries that HF/PyTorch preserves.

This was discovered while working on Qwen3 LoRA support in #173, where base-model numerical drift had to be separated from LoRA-specific correctness issues.

Tracked repro:

- Model: Qwen3-4B local weights matching the public Qwen3-4B release
- Prompt: `Tell me a story`
- Prompt token ids: `[40451, 752, 264, 3364]`
- Truth source: HF `transformers` incremental decode on CUDA, `torch.bfloat16`

The original visible mismatch happened at generated step 5:

| Step | HF GPU bf16 | PegaInfer before fix |
| ---: | --- | --- |
| 5 | token `6941` (`" named"`) | token `879` (`" who"`) |

HF had a real one-bucket bf16 margin:

| Candidate | HF raw bf16 logit | PegaInfer before fix |
| --- | ---: | ---: |
| `6941` (`" named"`) | `20.75` | tied |
| `879` (`" who"`) | `20.625` | tied |

## Root Cause

This was not a LoRA route issue, scheduler issue, CUDA Graph issue, lm_head issue, or sampler-only issue. The tracked mismatch came from numerical policy differences in the Qwen3 forward path.

The accepted fix set has four parts:

1. Preserve bf16 materialization on residual add before RMSNorm.
2. Match HF/PyTorch greedy tie behavior by selecting the lowest token id on exact logits ties.
3. Split Qwen3 MLP `gate_proj` and `up_proj` GEMMs instead of using one stacked `gate_up_proj` GEMM.
4. Match HF/PyTorch RoPE elementwise bf16 materialization inside fused q/k norm+RoPE.

The final remaining root cause was the RoPE arithmetic boundary. PegaInfer fused the RoPE multiply/add as:

```text
result = bf16(lo * cos - hi * sin)
result = bf16(lo * sin + hi * cos)
```

HF/PyTorch bf16 eager behavior materializes each elementwise multiply first:

```text
lo_cos = bf16(lo * cos)
hi_sin = bf16(hi * sin)
result = bf16(lo_cos - hi_sin)

lo_sin = bf16(lo * sin)
hi_cos = bf16(hi * cos)
result = bf16(lo_sin + hi_cos)
```

That missing materialization was enough to move later logits across bf16 rounding boundaries.

## Evidence

After the residual-add, tie-break, and split-MLP fixes, the first remaining divergence moved to generated step 48:

| Candidate | HF raw bf16 logit | PegaInfer raw bf16 logit |
| --- | ---: | ---: |
| `6801` (`" wants"`) | `17.375` | `17.375` |
| `76261` (`" imaginative"`) | `17.375` | `17.5` |
| `702` (`" has"`) | `17.25` | `17.25` |

HF selected lower id `6801` from the tie; PegaInfer selected `76261` because its fused path raised that candidate by one bf16 bucket.

The final split isolated q/k norm+RoPE:

- HF captured `q_norm_rope` exactly equals a PyTorch bf16 elementwise RoPE recomputation: mean abs `0.0`, max abs `0.0`.
- HF GPU cos/sin and PegaInfer CPU-generated bf16 cos/sin were identical at the checked position: mean abs `0.0`, max abs `0.0`.
- PegaInfer `q_norm_rope` still differed before the RoPE fix with identical q input and identical cos/sin values.

Therefore the issue was not the RoPE cache. It was the fused RoPE arithmetic boundary.

## Fix Set

Code paths changed:

- `pegainfer-qwen3-4b/src/prefill.rs`
- `pegainfer-qwen3-4b/src/batch_decode.rs`
- `pegainfer-qwen3-4b/src/batch_decode_dag.rs`
- `pegainfer-qwen3-4b/src/unified_forward.rs`
- `pegainfer-kernels/src/ops/sampling.rs`
- `pegainfer-kernels/csrc/prefill_attention.cu`

Key behavior changes:

- Use `fused_add_rms_norm_round_batch_into` for Qwen3 residual-add + RMSNorm paths that need HF-style bf16 residual materialization.
- Route greedy sampling through local `argmax_cuda` so exact ties choose the lower token id, matching `torch.argmax`.
- Run MLP gate/up projections as two row-sliced GEMMs, matching HF's separate `gate_proj` and `up_proj` module boundaries.
- Round each RoPE multiply to bf16 before the final add/subtract in `prefill_qk_norm_rope_kernel`.

## Negative Experiments

These did not fix the remaining drift and should not be treated as root cause:

- Changing shared dense GEMM from `CUBLAS_GEMM_DEFAULT_TENSOR_OP` to `CUBLAS_GEMM_DEFAULT`: first diff stayed at step 48.
- HF `DISABLE_ADDMM_CUDA_LT=1`: HF output unchanged through 128 tokens; still differed from pre-final-fix PegaInfer at step 48.
- Global `CUBLAS_COMPUTE_32F_PEDANTIC`: regressed the original step-5 mismatch.
- Decode workspace-backed cuBLAS with CUDA Graph disabled: first diff stayed at step 48.
- RMSNorm normalized-bf16-before-weight experiment: changed the step-48 failure but did not align greedy output.
- q/k serial reduction experiment: aligned one forced step-48 candidate check but regressed free generation to step 5.
- HF attention backend control: HF SDPA/default and HF eager agreed through step 61, so backend choice did not explain PegaInfer's step-48 drift.

## Validation

Local validation:

```bash
PEGAINFER_CUDA_SM=80 cargo check --release -p pegainfer-qwen3-4b --bin qwen3_decode_probe --bin qwen3_greedy_trace --bin qwen3_linear_probe
python3 -m py_compile \
  tools/accuracy/hf_dump_qwen3_greedy.py \
  tools/accuracy/compare_qwen3_final_hidden.py \
  tools/accuracy/compare_qwen3_prefill_kv.py
```

GPU worker validation:

```bash
cargo fmt --all --check
cargo check --release -p pegainfer-qwen3-4b --bin qwen3_greedy_trace --bin qwen3_decode_probe --bin qwen3_linear_probe
```

HF/PegaInfer greedy trace comparison:

- 128 generated tokens: no first diff.
- 256 generated tokens: no first diff.
- Last 16 generated tokens in the 256-token run matched exactly:
  `[374, 1602, 10838, 323, 6801, 311, 1477, 264, 1616, 311, 6177, 13, 2932, 374, 1602, 11521]`

Evidence artifact names from the GPU worker:

- `target/accuracy/peg_qwen3_tell_story_128_final_rope.json`
- `target/accuracy/hf_qwen3_tell_story_256_final_rope.json`
- `target/accuracy/peg_qwen3_tell_story_256_final_rope.json`

## Follow-up

Add a regression gate for this exact Qwen3-4B base HF GPU bf16 greedy trace. The existing Qwen3 e2e baseline appears to encode the old drift branch and should be regenerated or replaced only after deciding the desired baseline policy.

**Comment by NolanHo** [comment]
Opened PR #175 for the Qwen3-4B base numerical fixes. This drift was found while separating base-model correctness from Qwen3 LoRA support work in #173.

**Comment by xiaguan** [comment]
should we close this? since #175  is merged

**Comment by NolanHo** [comment]
Yes, I think we can close this now.
